### PR TITLE
Allow setting external contest source with contest.yaml

### PR DESCRIPTION
**Pull request by vmcj**
This makes it easier to setup the analyst instance during the finals.

I've done it so that we only overwrite the fields if explicit set so it is possible to remove the password from the contest and not remove it when someone updates the contest again to alter the time for example.

Closes: https://github.com/DOMjudge/domjudge/issues/2179

### CODE REVIEW

File: webapp/src/Service/ImportExportService.php [hunk 1]
```diff
@@ -251,6 +252,23 @@ public function importContestData(mixed $data, ?string &$errorMessage = null, st
         }
 
         $this->em->persist($contest);
+
+        $shadow_type = $data['shadow_type'] ?? null;
```
Comment (nickygerritsen) [inline webapp/src/Service/ImportExportService.php:256]: We should document that you can do this whole thing

Comment (vmcj) [reply webapp/src/Service/ImportExportService.php:256]: Makes sense, although the only users using this are us IMO.

Comment (nickygerritsen) [reply webapp/src/Service/ImportExportService.php:256]: Which I agree with but I don't want to check the code to see how the fields are called. Also, maybe we use a subkey?
```yaml
shadow:
  type: xxx
```
etc?

Also makes the code more readable.

File: doc/manual/shadow.rst [hunk 2]
```diff
@@ -54,6 +54,20 @@ In the DOMjudge admin interface, go to the *External Contest Sources* page and
...
+  shadow:
+    type: contest-api
+    source: http(s)://url_to_external_system/api/contests/<contest_id>
+    username: user_in_external_system
+    password: password_for_user
+
+or::
+
+  shadow:
+    type: contest-api
```
Comment (nickygerritsen) [inline doc/manual/shadow.rst:68]:
```suggestion
    type: contest-archive
```

right?

File: webapp/src/Service/ImportExportService.php [hunk 3]
```diff
@@ -251,6 +252,22 @@ public function importContestData(mixed $data, ?string &$errorMessage = null, st
...
+            $externalSource = $this->em->getRepository(ExternalContestSource::class)->findOneBy(['contest' => $contest]) ?: new ExternalContestSource();
+            $externalSource->setContest($contest);
+            foreach (['source', 'username', 'password', 'type'] as $field) {
+                // Overwrite the existing value if the property is defined in the data: $externalSource-setSource($data['shadow']['source'])
+                $fieldFunc = 'set'.ucwords($field);
+                $fieldArgs = [$shadow[$field]];
+                if (isset($shadow[$field])) {
+                    $externalSource->$fieldFunc(...$fieldArgs);
+                }
+            }
```
Comment (nickygerritsen) [inline webapp/src/Service/ImportExportService.php:267]:
```suggestion
            foreach ($shadow as $field => $value) {
                $fieldFunc = 'set'.ucwords($field);
                $fieldArgs = [$value];
                if (method_exists($externalSource, $fieldFunc)) {
                    $externalSource->$fieldFunc(...$fieldArgs);
                }
            }
```

would now also work right? That way if we ever add fields it just works ™️

**Review by nickygerritsen — APPROVED** [review]

File: doc/manual/shadow.rst (hunk 2 above)
Comment (vmcj) [reply doc/manual/shadow.rst]: yes, wrong copy paste.

**Pull request merged**